### PR TITLE
Add node class

### PR DESCRIPTION
**Pull request by vhaller**
All recent changes merged to one pull request.

### CODE REVIEW

File: src/Node.lua [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Node.__index = Node
+
+function Node:new()
+  return setmetatable({parent = nil, children = {}}, Node)
+end
+
+function Node:add_child(child)
+  self.children[#self.children + 1] = child
+  child.parent = self
+  child.index  = #self.children
```
Comment (kulla) [inline src/Node.lua]: <del>Why do you set the `index` attribute to the children's index? Is this attribute used somewhere?!</del>


Delete this line. It's really ugly code :smile:

File: test/Node_test.lua [hunk 2]
```diff
@@ -0,0 +1,39 @@
+luaunit = require("luaunit")
+Node = require("Node")
+
+local TestNode = {}
+
+local root, node_1, node_2, node_3 = Node:new(), Node:new(), Node:new(), Node:new()
+root:add_child(node_1)
+root:add_child(node_2)
+node_2:add_child(node_3)
```
Comment (kulla) [inline test/Node_test.lua]: Use a `setUp()` function here (http://luaunit.readthedocs.io/en/latest/#grouping-tests-setup-teardown-functionality )

File: test/Node_test.lua [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+function TestNode:test_new()
+  luaunit.assertTrue(getmetatable(root) == Node)
```
Comment (kulla) [inline test/Node_test.lua]: Use [assertIs](http://luaunit.readthedocs.io/en/latest/#assertIs):

```
luaunit.assertIs(getmetatable(root), Node)
```

File: test/Node_test.lua [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  luaunit.assertNotNil(root)
+  luaunit.assertNil(root.parent)
```
Comment (kulla) [inline test/Node_test.lua]: Missing:

```
luaunit.assertItemsEquals(root.children, {})
```

See http://luaunit.readthedocs.io/en/latest/#assertItemsEquals

File: test/Node_test.lua [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+end
+
+function TestNode:test_add()
+  luaunit.assertTrue(node_1.parent == root)
+  luaunit.assertTrue(node_2.parent == root)
+  luaunit.assertTrue(node_3.parent == node_2)
+  luaunit.assertTrue(#root.children == 2 and #node_2.children == 1)
```
Comment (kulla) [inline test/Node_test.lua]: Use `assertEquals` and write it in two lines...

File: src/Node.lua (hunk 1 above)
Comment (vhaller) [reply src/Node.lua]: Right! I didn't notice.

**Comment by kulla** [comment]
@vhaller: Have a look on [how to format a git commit message](http://chris.beams.io/posts/git-commit/) :smile: However, good work so far!

File: test/Node_test.lua [hunk 6]
```diff
@@ -36,4 +36,15 @@ function TestNode:test_level()
   luaunit.assertTrue(node_3:level() == 2)
 end
 
+function TestNode:test_find()
+  luaunit.assertTrue(root:find(function(node) return node == node_3 end) == node_3)
```
Comment (kulla) [inline test/Node_test.lua]: Use [`assertIs`](http://luaunit.readthedocs.io/en/latest/#assertIs) here... (and in the following lines)

File: test/Node_test.lua [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+  luaunit.assertEquals(#root.children, 2)
+  luaunit.assertEquals(#node_2.children, 1)
+end
+
+function TestNode:test_last_child()
+  luaunit.assertIs(root:last_child(), node_2)
+  luaunit.assertIs(node_2:last_child(), node_3)
+
+  luaunit.assertFalse(node_1:last_child())
```
Comment (kulla) [inline test/Node_test.lua]: Use `assertIsNil` here...

File: test/Node_test.lua [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+root:add_child(node_1)
+root:add_child(node_2)
+node_2:add_child(node_3)
+
+function TestNode:test_new()
+  luaunit.assertIs(getmetatable(root), Node)
+
+  luaunit.assertNotNil(root)
+
+  luaunit.assertNil(root.parent)
```
Comment (kulla) [inline test/Node_test.lua]: <del>`luaunit.assertItemsEquals(root.children, {})` is missing</del>

File: test/Node_test.lua [hunk 9]
```diff
@@ -0,0 +1,56 @@
+luaunit = require("luaunit")
+Node = require("Node")
+
+local TestNode = {}
+
+local root, node_1, node_2, node_3 = Node:new(), Node:new(), Node:new(), Node:new()
+root:add_child(node_1)
+root:add_child(node_2)
+node_2:add_child(node_3)
+
```
Comment (kulla) [inline test/Node_test.lua:12]: <del>Use a `setUp()` function here (http://luaunit.readthedocs.io/en/latest/#grouping-tests-setup-teardown-functionality )</del>

File: test/Node_test.lua [hunk 10]
```diff
@@ -0,0 +1,56 @@
+luaunit = require("luaunit")
+Node = require("Node")
+
+local TestNode = {}
+
+local root, node_1, node_2, node_3 = Node:new(), Node:new(), Node:new(), Node:new()
```
Comment (kulla) [inline test/Node_test.lua]: <del>use variable names `node1`, `node2`, `node3`</del>

File: test/Node_test.lua [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  luaunit.assertNotNil(root)
+
+  luaunit.assertNil(root.parent)
+end
+
+function TestNode:test_add()
+  luaunit.assertIs(node_1.parent, root)
+  luaunit.assertIs(node_2.parent, root)
+  luaunit.assertIs(node_3.parent, node_2)
+  luaunit.assertItemsEquals(root.children, {node_1, node_2})
```
Comment (kulla) [inline test/Node_test.lua]: <del>Add

```
luaunit.assertItemsEquals(node_1.children, {})
luaunit.assertItemsEquals(node_2.children, { node_3 })
luaunit.assertItemsEquals(node_3.children, {})</del>
```

File: test/Node_test.lua [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+  luaunit.assertNotNil(root)
+
+  luaunit.assertNil(root.parent)
+end
+
+function TestNode:test_add()
+  luaunit.assertIs(node_1.parent, root)
+  luaunit.assertIs(node_2.parent, root)
+  luaunit.assertIs(node_3.parent, node_2)
```
Comment (kulla) [inline test/Node_test.lua]: eine Leerzeile hier (ich weiß, dass ist ne ziemliche Kleinigkeit)... :smile:

File: test/Node_test.lua [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  luaunit.assertItemsEquals(root.children, {node_1, node_2})
+
+  luaunit.assertEquals(#root.children, 2)
+  luaunit.assertEquals(#node_2.children, 1)
```
Comment (kulla) [inline test/Node_test.lua]: Diese zwei Zeilen können gelöscht werden, da sie bereits durch

```
luaunit.assertItemsEquals(root.children, {node_1, node_2})
```

abgedeckt werden...

File: test/Node_test.lua [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+function TestNode:test_new()
+  luaunit.assertIs(getmetatable(root), Node)
+
+  luaunit.assertNotNil(root)
+
+  luaunit.assertNil(root.parent)
+end
+
+function TestNode:test_add()
```
Comment (kulla) [inline test/Node_test.lua:24]: <del>Add `luaunit.assertIsNil(root.parent)`</del>

File: test/Node_test.lua [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  luaunit.assertEquals(node_1:level(), 1)
+  luaunit.assertEquals(node_2:level(), 1)
+  luaunit.assertEquals(node_3:level(), 2)
+end
+
+function TestNode:test_find()
+  luaunit.assertIs(root:find(function(node) return node == node_3 end), node_3)
+  luaunit.assertIs(root:find(function(node) return node.parent ~= nil end), node_1)
+
+  luaunit.assertNotIs(root:find(function(node) return node:level() == 1 end), node_2)
```
Comment (kulla) [inline test/Node_test.lua]: <del>Mach hier ein `node:level() == 2` raus (muss dann `node_2` zurückgeben)</del>

**Review by kulla — CHANGES_REQUESTED** [review]
Last things which are needed before the merge...

**Comment by kulla** [comment]
At the end only one function is missing:
- [x] `Node:find_parent()`

File: test/Node_test.lua (hunk 8 above)
Comment (vhaller) [reply test/Node_test.lua]: When I add this assertion, the test fails:

## Failed tests:

1) Test_Node.test_new
.\test\Node_test.lua:18: Contents of the tables are not identical:
Expected:
{}
Actual:
<table: 00630460> {
    <table: 00630488> {
        children=<table: 006304B0> {},
        parent=<table: 00630438> {children=<table: 00630460>}
    },
    <table: 006304D8> {
        children=<table: 00630500> {<table: 00630528> {children=<table: 00630550> {}, parent=<table: 006304D8>}},
        parent=<table: 00630438>
    }
}
stack traceback:
        .\test\Node_test.lua:18: in function 'Test_Node.test_new'

File: test/Node_test.lua (hunk 14 above)
Comment (vhaller) [inline test/Node_test.lua]: To TestNode:test_new() or to TestNode:test_add()?

Comment (vhaller) [reply test/Node_test.lua:24]: Isn't it there already?

File: test/Node_test.lua (hunk 15 above)
Comment (vhaller) [reply test/Node_test.lua]: In Zeile 48?

File: test/Node_test.lua (hunk 14 above)
Comment (kulla) [reply test/Node_test.lua:24]: `test_add` hasn't this test already... (each test case needs all assertations in order have a full coverage even if there are overlaps between tests)

File: test/Node_test.lua (hunk 8 above)
Comment (kulla) [reply test/Node_test.lua]: <del>Stimmt. Add `local root = Node:new()` at the beginning of the test...</del>

File: test/Node_test.lua (hunk 15 above)
Comment (kulla) [inline test/Node_test.lua]: ah, ne, passt hier...

**Review by kulla — COMMENTED** [review]
AW

File: test/Node_test.lua (hunk 15 above)
Comment (kulla) [inline test/Node_test.lua]: Passt

**Review by kulla — APPROVED** [review]

File: test/Node_test.lua (hunk 15 above)
Comment (kulla) [reply test/Node_test.lua]: Passt

**Review by kulla — CHANGES_REQUESTED** [review]

File: test/Node_test.lua (hunk 8 above)
Comment (vhaller) [reply test/Node_test.lua]: Alright, thanks.

File: test/Node_test.lua (hunk 14 above)
Comment (vhaller) [reply test/Node_test.lua:24]: I understand.

**Review by kulla — APPROVED** [review]

**Pull request merged**